feat(plugins): Support configurable extensions

robzienert opened the pull request:
PF4J has two concepts of extension: System extensions and Plugin-provided extensions. This PR refactors the current codebase to support configuring both concepts in a common way.

An example config might look something like this:

```yaml
spinnaker:
  plugins:
    robzienert.helloworld:
      enabled: true
      extensions:
        robzienert.helloworld-stage1:
          config:
            defaultRecipient: "Kork Reviewers"
        robzienert.helloworld-stage2:
          config:
            defaultRecipient: "Rob"
  extensions:
    robzienert.something-cool:
      enabled: true
      config:
        myConfig: myValue
```

The extension `$namespace.$id` is defined on the extension and is used to resolve config, the same logic (just with a different pointer) to load extension config:

```kt
@SpinnakerExtension(namespace = "robzienert", id = "something-cool")
class SomethingCool : SomeSpinnakerExtensionPoint, ConfigurableExtension<SomethingCool.CoolConfig> {
  lateinit var config: CoolConfig

  fun setConfiguration(config: CoolConfig) {
    this.config = config
  }
}

class CoolConfig(
  val myConfig: String
)
```

A `ConfigResolver` is used to find configs and the default implementation is `SpringEnvironmentExtensionConfigResolver`. This resolver will use the parent service Spring Environment `PropertySources` for all configs. Sourcing from other places, like Front50, would be a matter of overriding the `ConfigResolver` - perhaps this becomes a resolver chain in the future.

I had to refactor quite a lot of things to make this PR a thing, but I've moved stuff around and I feel the codebase is easier to reason about as a result. One departure from stock PF4J is that all extensions are initialized once as singletons.

cfieber commented on kork-plugins-api/src/main/java/com/netflix/spinnaker/kork/plugins/ConfigurableExtension.java:
```diff
@@ -0,0 +1,25 @@
+
+/** Marks an Extension as being capable of accepting a configuration. */
+@Alpha
+public interface ConfigurableExtension {
```
I also wonder in that case if there is kotlin reflection business to derive the generic type from the extension rather that requiring it on the annotation? (I don't know the reflection stuff in kotlin super well)

cfieber commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/config/SpringEnvironmentExtensionConfigResolver.kt:
```diff
@@ -0,0 +1,100 @@
+    .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+    .findAndRegisterModules()
+
+  override fun <T> resolve(coordinates: ConfigCoordinates, expectedType: Class<T>): T {
```
what happens if you try and let the spring binding stuff in the environment do the heavy lifting here?

`environment.getProperty("spinnaker.plugins.${coordinates.pluginNamespace}.${coordinates.pluginId}.extensions.${coordinates.extensionNamespace}.${coordinates.extensionId}.config", expectedType)`

robzienert replied:
Oh haha. Right.

robzienert replied:
For readers: Spring's PropertySources do not resolve to a non-primitive type. I wound up having to fold the enumerable property sources into a single `Properties` object, then use Jackson to read that value into a map, and finally use Jackson to cast the config at `pointer` to the desired config type.

jonsie commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpringExtensionFactory.kt:
```diff
@@ -26,20 +30,24 @@ import org.springframework.core.env.StandardEnvironment
       ?: throw PluginRuntimeException("Failed to create object of extension class: $extensionClass")
-    val pluginWrapper = pluginManager.whichPlugin(extensionClass) ?: return extension
 
+    val anno = getSpinnakerExtensionAnno(extension)
```
I prefer to abbreviate `annotation` as `annot`.

christopherthielen commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpringExtensionFactory.kt:
```diff
@@ -50,16 +58,72 @@ open class SpringExtensionFactory(
+   */
+  private fun loadConfig(extension: Any, anno: SpinnakerExtension, coordinates: ConfigCoordinates) {
+    if (anno.config.java == MissingConfiguration::class.java) {
+      return
```
I had assumed that `MissingConfiguration` was an error state based on the name of the class.

srekapalli commented on kork-plugins-api/src/main/java/com/netflix/spinnaker/kork/plugins/MissingConfiguration.java:
```diff
@@ -0,0 +1,24 @@
+ * configuration.
+ *
+ * @see SpinnakerExtension
+ */
```
EmptyConfiguration or NoConfiguration ... Name looks like an Exception condition or something.

marchello2000 commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionBeanDefinitionRegistryPostProcessor.kt:
```diff
@@ -26,28 +27,51 @@ import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProce
+
+        beanFactory.registerSingleton(beanName, bean)
+      } catch (e: ClassNotFoundException) {
+        throw IntegrationException("Could not find system extension class '$it'", e)
```
is it the intent that a failure of one extension fails all extensions/service start?
(also, deliberate that other exceptions aren't caught?  I assume that if the extension throws in the .ctor/somewhere on the init path) it would also fail all other extensions from loading

robzienert replied:
I'm going to re-arrange the class so the try-catch is specifically surrounding the class loading instead of the whole logic (as it should be).

marchello2000 commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpringExtensionFactory.kt:
```diff
@@ -50,16 +58,72 @@ open class SpringExtensionFactory(
+   * Extensions within Spinnaker MUST use the [SpinnakerExtension] annotation, rather than the default PF4J
+   * [org.pf4j.Extension] annotation.
+   */
+  private fun getSpinnakerExtensionAnno(extension: Any): SpinnakerExtension =
```
What @jonsie said `Anno` -> `Annotation`

srekapalli commented on kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionBeanDefinitionRegistryPostProcessor.kt:
```diff
@@ -26,28 +27,51 @@ import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProce
+        val extensionClass = javaClass.classLoader.loadClass(it)
+        val bean = pluginManager.extensionFactory.create(extensionClass)
+        val beanName = "${extensionClass.simpleName.decapitalize()}SystemExtension"
+
```
Will it be useful to publish an Spring ApplicationEvent here , which allow us to possibly do some reporting/telemetry ?

robzienert replied:
Great thought!

robzienert commented:
Updated from feedback:

1. `s/anno/annot`. Also changed the `Anno` abbreviated method to `Annotation`.
1. Removed `config` from `@SpinnakerExtension` and made `ConfigurableExtension<T>`. The `setConfiguration(T)` method needed to be called through reflection, so `SpringExtensionFactory` got a little more gross, but the end-user developer experience will be much more pleasant now. This change also means that `MissingConfiguration` no longer exists.
1. Added an `ExtensionLoaded` Spring `ApplicationEvent` that is emitted for every extension loaded.
1. Refactored `SpringEnvironmentExtensionConfigResolver` to use the `Environment.getProperty(String, Class<?>)` method, which has greatly simplified this class.

ajordens commented:
Latest cut is looking pretty clean to me, nice work.
